### PR TITLE
Add conformance test for combination of extended match types with core types

### PR DESCRIPTION
**Pull request by gauravkghildiyal**
/kind test
/area conformance

**What type of PR is this?**

**What this PR does / why we need it**:
Add conformance test for combination of extended match types with core types

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes-sigs/gateway-api/issues/1962

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: conformance/tests/httproute-matching-combinations.go
Comment (sunjayBhatia) [inline conformance/tests/httproute-matching-combinations.go:113]: seems like we need a test case that has all the expected content that would match a particular rule except the method to make sure all the match criteria is respected, I don't think there is any rule that matches just `/` (w/ no header or query params) so might be good to add another test case that e.g. sends a GET to `/path3?animal=whale` (I don't think the above request to `/?animal=whale` satisfies this b/c I don't think there is a rule that expects to match on that with a different method)

Comment (gauravkghildiyal) [reply conformance/tests/httproute-matching-combinations.go:113]: Sure I've added that. We do have cases which "matched all except one" but none where the "one" which didn't match was an HTTP method.

**Review by sunjayBhatia — COMMENTED** [review]
test passes w/ Contour main 👍🏽

**Review by arkodg — APPROVED** [review]
LGTM thanks !

File: conformance/tests/httproute-matching-combinations.yaml
Comment (mlavacca) [inline conformance/tests/httproute-matching-combinations.yaml:9]: Thank you @gauravkghildiyal! WDYT about adding another rule with all the possible `matches` set (path, header, method, queryparam)? If I see it correctly, we have many combinations, but none of them specifies all the possible `match` types.

Comment (gauravkghildiyal) [reply conformance/tests/httproute-matching-combinations.yaml:9]: Sure thing! Done. Thanks for the review

File: conformance/tests/httproute-matching-combinations.yaml
Comment (robscott) [inline conformance/tests/httproute-matching-combinations.yaml:4]: I think it would be very helpful to have some tests covering a use case where multiple rules match the same request. For example

Rule A:
- Path match on /foo

Rule B:
- Query param match on animal=whale

Test: Request to `/foo?animal=whale`. 

You may already have some similar test cases in here that I'm missing, but if not, I think it would be really helpful to have something like this. This is conceptually similar to https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/httproute-matching-across-routes.go, but would cover extended concepts like query param and method matching.

File: conformance/tests/httproute-matching-combinations.go
Comment (robscott) [inline conformance/tests/httproute-matching-combinations.go:42]: Similar to how the Redirect tests were split up into separate tests for each extended feature combined with all core features, it could be helpful to have subsets of this available. For the record, I don't think that needs to block this PR from merging, but could be a good follow up.

**Review by robscott — COMMENTED** [review]
Thanks @gauravkghildiyal! I've got a couple non-blocking comments that don't need to block this PR, but could make for good follow ups here or in a future PR.

/approve

File: conformance/tests/httproute-matching-combinations.yaml
Comment (gauravkghildiyal) [reply conformance/tests/httproute-matching-combinations.yaml]: Interesting. I completely missed this. Thanks Rob! I've added some to get started

File: conformance/tests/httproute-matching-combinations.go
Comment (gauravkghildiyal) [reply conformance/tests/httproute-matching-combinations.go]: I went ahead and split it up here itself, since we're anyway waiting for LGTM. I wasn't aware we try to maintain such a pattern. But yes it does make sense. 

@robscott -- I guess this implies that if one wishes to add tests that cover ALL match types, that would go in it's own file/test. Will see if that is needed in a separate PR.

**Review by gauravkghildiyal — COMMENTED** [review]
Thanks for the review Rob!

File: conformance/tests/httproute-matching-combinations.go
Comment (robscott) [reply conformance/tests/httproute-matching-combinations.go]: Yep, I think separate files are _probably_ sufficient here, but don't feel too strongly either way.

**Comment by robscott** [comment]
Thanks @gauravkghildiyal!

/lgtm
/cherry-pick release-0.7

File: conformance/tests/httproute-query-param-matching.go
Comment (sunjayBhatia) [inline conformance/tests/httproute-query-param-matching.go:143]: Looks like this test case added after my review doesn't work with Contour

this assumes header matching gets precedence over method matching, but I'm not sure that is expressed anywhere explicitly in the spec, see https://github.com/kubernetes-sigs/gateway-api/issues/1993

Comment (gauravkghildiyal) [reply conformance/tests/httproute-query-param-matching.go:143]: Yes that was how I interpreted it -- given method matching is not in the list (and the only thing not in the list), it must have the least priority.

I'm sorry for causing any issues. I've sent https://github.com/kubernetes-sigs/gateway-api/pull/2052 as a temporary solution to help unblock (if this is the path we wish to take)

Comment (sunjayBhatia) [reply conformance/tests/httproute-query-param-matching.go:143]: thanks! yeah not a super strong opinion on where it lands in terms of priority, but should probably be something explicitly debated/documented 👍🏽

**Pull request merged**
